falcon/cyutil/uri.pyx

```diff
@@ -252,14 +266,159 @@ cdef cy_parse_query_string(unsigned char* data, Py_ssize_t length,
     return result
 
 
+cdef Py_ssize_t _cy_probe_encoded_length(unsigned char* data, Py_ssize_t length,
+                                         bint is_value, bint check_is_escaped):
+    cdef Py_ssize_t pos
+    cdef Py_ssize_t extra_length = 0
+    cdef bint already_escaped = check_is_escaped
+
+    # PERF(vytas): Here we unroll the code for different values of is_value.
+    #   This could by DRYed at a minor perf cost, but, OTOH, these
+    #   redundant blocks are not too large anyway.
+    if is_value:
+        for pos in range(length):
+            # NOTE(vytas): See _UNRESERVED in uri.py.
+            if data[pos] not in (b'ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz'
+                                 b'0123456789-._~'):
```

Comment:
are we sure that a set is not faster? in python it is significantly faster

need to use `ord(data[pos])` in case

Reply:
same in the other places. it would also simplify drying this

Reply:
Yeah, I also read about it before how Cython can optimize this kind of inclusion test.

Cf
```cython
HEX_CHARS = frozenset(b'0123456789ABCDEFabcdef')


def count_chars_in1():
    cdef unsigned char c
    cdef int result = 0

    for c in range(0xFF):
        if c in b'0123456789ABCDEFabcdef':
            result += 1

    return result


def count_chars_in2():
    cdef unsigned char c
    cdef int result = 0

    for c in range(0xFF):
        if c in HEX_CHARS:
            result += 1

    return result
```

You wouldn't get this result in pure Python, but when cythonized, the first function is ~14x faster.

Reply:
hanks for checking